zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/adhocsubprocess/parameter/AdHocActivityParameterExtractor.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.engine.processing.adhocsubprocess.parameter;
+
+import io.camunda.zeebe.engine.processing.adhocsubprocess.AdHocActivityMetadata.AdHocActivityParameter;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import org.camunda.feel.syntaxtree.FunctionInvocation;
+import org.camunda.feel.syntaxtree.ParsedExpression;
+import scala.Product;
+import scala.jdk.javaapi.CollectionConverters;
+
+/**
+ * Generic parameter extractor delegating actual function call handling to individual
+ * FunctionCallParameterExtractors.
+ */
+public class AdHocActivityParameterExtractor {
+
+  private final Map<String, FeelFunctionCallParameterExtractor> parameterExtractors =
+      new LinkedHashMap<>();
+
+  public AdHocActivityParameterExtractor() {
+    registerParameterExtractor(new FromAiFeelFunctionCallParameterExtractor());
+  }
+
+  private void registerParameterExtractor(final FeelFunctionCallParameterExtractor extractor) {
+    parameterExtractors.put(extractor.functionName(), extractor);
+  }
+
+  public List<AdHocActivityParameter> extractParameters(final ParsedExpression parsedExpression) {
```

Comment:
This is the entry point to iterate over an expression and to extract all supported function calls down the tree. Actual support for function invocations is delegated to `FeelFunctionCallParameterExtractor`, so down the line we could support additional function instead only `fromAi()` (e.g. less AI specific such as `parameter()` or `generated()`).